File: redis_metadata_parser.c
/**
 * Redis Metadata Buffer Parser - C Implementation
 * 
 * This library provides functions to parse Redis response buffers that contain
 * both metadata and data when the read-response-metadata feature is enabled.
 */

#include "redis_metadata_parser.h"
#include <string.h>
#include <stdlib.h>
#include <arpa/inet.h>
#include <endian.h>

#define MAGIC_HEADER "RDMT"
#define MAGIC_SIZE 4

/**
 * Initialize a metadata structure
 */
void redis_metadata_init(redis_metadata_t *metadata) {
    if (!metadata) return;
    
    memset(metadata, 0, sizeof(redis_metadata_t));
}

/**
 * Free resources in a metadata structure
 */
void redis_metadata_free(redis_metadata_t *metadata) {
    if (!metadata) return;
    
    if (metadata->key) {
        free(metadata->key);
        metadata->key = NULL;
    }
    if (metadata->field) {
        free(metadata->field);
        metadata->field = NULL;
    }
}

/**
 * Initialize a response structure
 */
void redis_response_init(redis_response_t *response) {
    if (!response) return;
    
    memset(response, 0, sizeof(redis_response_t));
    redis_metadata_init(&response->metadata);
}

/**
 * Free resources in a response structure
 */
void redis_response_free(redis_response_t *response) {
    if (!response) return;
    
    redis_metadata_free(&response->metadata);
    if (response->data) {
        free(response->data);
        response->data = NULL;
    }
    response->data_size = 0;
}

/**
 * Get human-readable data type name
 */
const char* redis_data_type_name(uint8_t data_type) {
    switch (data_type) {
        case REDIS_TYPE_STRING: return "string";
        case REDIS_TYPE_HASH:   return "hash";
        case REDIS_TYPE_LIST:   return "list";
        case REDIS_TYPE_SET:    return "set";
        case REDIS_TYPE_ZSET:   return "zset";
        case REDIS_TYPE_STREAM: return "stream";
        default:                return "unknown";
    }
}

/**
 * Parse a Redis metadata buffer
 * 
 * @param buffer Raw buffer from Redis
 * @param buffer_size Size of the buffer
 * @param response Output response structure (caller must free with redis_response_free)
 * @return REDIS_PARSE_OK on success, error code on failure
 */
redis_parse_result_t redis_parse_metadata_buffer(
    const uint8_t *buffer,
    size_t buffer_size,
    redis_response_t *response
) {
    if (!buffer || !response || buffer_size < MAGIC_SIZE) {
        return REDIS_PARSE_INVALID_ARGS;
    }
    
    // Initialize response
    redis_response_init(response);
    
    // Check magic header
    if (memcmp(buffer, MAGIC_HEADER, MAGIC_SIZE) != 0) {
        return REDIS_PARSE_NO_METADATA;
    }
    
    if (buffer_size < 20) { // Minimum header size
        return REDIS_PARSE_BUFFER_TOO_SMALL;
    }
    
    const uint8_t *ptr = buffer;
    
    // Skip magic header
    ptr += MAGIC_SIZE;
    
    // Parse header fields
    uint32_t total_size = ntohl(*(uint32_t*)ptr);
    ptr += 4;
    
    uint32_t metadata_size = ntohl(*(uint32_t*)ptr);
    ptr += 4;
    
    uint32_t data_size = ntohl(*(uint32_t*)ptr);
    ptr += 4;
    
    // Validate buffer size
    if (total_size != buffer_size) {
        return REDIS_PARSE_SIZE_MISMATCH;
    }
    
    // Parse key length
    if (ptr + 4 > buffer + buffer_size) {
        return REDIS_PARSE_BUFFER_TRUNCATED;
    }
    uint32_t key_len = ntohl(*(uint32_t*)ptr);
    ptr += 4;
    
    // Check if we have field length (hash format)
    uint32_t field_len = 0;
    size_t remaining_before_strings = metadata_size - (ptr - buffer);
    size_t min_remaining = 4 + 8 + 8 + 8 + 1; // data_len + timestamp + hits + misses + type
    
    if (remaining_before_strings >= min_remaining + 4) {
        if (ptr + 4 > buffer + buffer_size) {
            return REDIS_PARSE_BUFFER_TRUNCATED;
        }
        field_len = ntohl(*(uint32_t*)ptr);
        ptr += 4;
    }
    
    // Parse data length
    if (ptr + 4 > buffer + buffer_size) {
        return REDIS_PARSE_BUFFER_TRUNCATED;
    }
    uint32_t data_len = ntohl(*(uint32_t*)ptr);
    ptr += 4;
    
    if (data_len != data_size) {
        return REDIS_PARSE_SIZE_MISMATCH;
    }
    
    // Parse timestamp and stats
    if (ptr + 24 > buffer + buffer_size) {
        return REDIS_PARSE_BUFFER_TRUNCATED;
    }
    
    uint64_t access_timestamp = be64toh(*(uint64_t*)ptr);
    ptr += 8;
    
    uint64_t keyspace_hits = be64toh(*(uint64_t*)ptr);
    ptr += 8;
    
    uint64_t keyspace_misses = be64toh(*(uint64_t*)ptr);
    ptr += 8;
    
    // Parse data type
    if (ptr + 1 > buffer + buffer_size) {
        return REDIS_PARSE_BUFFER_TRUNCATED;
    }
    uint8_t data_type = *ptr;
    ptr += 1;
    
    // Parse key string
    if (ptr + key_len > buffer + buffer_size) {
        return REDIS_PARSE_BUFFER_TRUNCATED;
    }
    
    char *key = malloc(key_len + 1);
    if (!key) {
        return REDIS_PARSE_MEMORY_ERROR;
    }
    memcpy(key, ptr, key_len);
    key[key_len] = '\0';
    ptr += key_len;
    
    // Parse field string (if hash)
    char *field = NULL;
    if (field_len > 0) {
        if (ptr + field_len > buffer + buffer_size) {
            free(key);
            return REDIS_PARSE_BUFFER_TRUNCATED;
        }
        
        field = malloc(field_len + 1);
        if (!field) {
            free(key);
            return REDIS_PARSE_MEMORY_ERROR;
        }
        memcpy(field, ptr, field_len);
        field[field_len] = '\0';
        ptr += field_len;
    }
    
    // Parse actual data
    if (ptr + data_len > buffer + buffer_size) {
        free(key);
        if (field) free(field);
        return REDIS_PARSE_BUFFER_TRUNCATED;
    }
    
    uint8_t *data = malloc(data_len);
    if (!data) {
        free(key);
        if (field) free(field);
        return REDIS_PARSE_MEMORY_ERROR;
    }
    memcpy(data, ptr, data_len);
    
    // Fill response structure
    response->metadata.key = key;
    response->metadata.field = field;
    response->metadata.data_type = data_type;
    response->metadata.data_size = data_len;
    response->metadata.access_timestamp = access_timestamp;
    response->metadata.keyspace_hits = keyspace_hits;
    response->metadata.keyspace_misses = keyspace_misses;
    response->metadata.total_size = total_size;
    response->metadata.metadata_size = metadata_size;
    
    response->data = data;
    response->data_size = data_len;
    
    return REDIS_PARSE_OK;
}

/**
 * Get string representation of parse result
 */
const char* redis_parse_result_string(redis_parse_result_t result) {
    switch (result) {
        case REDIS_PARSE_OK:                return "Success";
        case REDIS_PARSE_NO_METADATA:      return "No metadata in buffer";
        case REDIS_PARSE_INVALID_ARGS:     return "Invalid arguments";
        case REDIS_PARSE_BUFFER_TOO_SMALL: return "Buffer too small";
        case REDIS_PARSE_BUFFER_TRUNCATED: return "Buffer truncated";
        case REDIS_PARSE_SIZE_MISMATCH:    return "Size mismatch";
        case REDIS_PARSE_MEMORY_ERROR:     return "Memory allocation error";
        default:                           return "Unknown error";
    }
} 

File: redis_metadata_parser.py
#!/usr/bin/env python3
"""
Redis Metadata Buffer Parser

This library provides functions to parse Redis response buffers that contain
both metadata and data when the read-response-metadata feature is enabled.

Buffer Format:
- Magic header "RDMT" (4 bytes)
- Total buffer size (4 bytes, network byte order)
- Metadata section size (4 bytes, network byte order)
- Data section size (4 bytes, network byte order)
- Key length (4 bytes, network byte order)
- Field length (4 bytes, network byte order) [hash only]
- Data length (4 bytes, network byte order)
- Access timestamp (8 bytes, network byte order)
- Keyspace hits (8 bytes, network byte order)
- Keyspace misses (8 bytes, network byte order)
- Data type (1 byte: 0=string, 1=hash, 2=list, etc.)
- Key string (variable length)
- Field string (variable length) [hash only]
- Actual data (variable length)
"""

import struct
import time
from typing import Dict, Any, Optional, Tuple, Union

class RedisMetadata:
    """Represents metadata from a Redis read operation."""
    
    def __init__(self):
        self.key: str = ""
        self.field: Optional[str] = None
        self.data_type: int = 0
        self.data_size: int = 0
        self.access_timestamp: int = 0
        self.keyspace_hits: int = 0
        self.keyspace_misses: int = 0
        self.total_size: int = 0
        self.metadata_size: int = 0

    @property
    def data_type_name(self) -> str:
        """Get human-readable data type name."""
        types = {
            0: "string",
            1: "hash", 
            2: "list",
            3: "set",
            4: "zset",
            5: "stream"
        }
        return types.get(self.data_type, "unknown")
    
    @property
    def access_time_formatted(self) -> str:
        """Get formatted access time."""
        return time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(self.access_timestamp))
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert metadata to dictionary."""
        result = {
            "key": self.key,
            "data_type": self.data_type_name,
            "data_size": self.data_size,
            "access_timestamp": self.access_timestamp,
            "access_time": self.access_time_formatted,
            "keyspace_hits": self.keyspace_hits,
            "keyspace_misses": self.keyspace_misses,
            "total_size": self.total_size,
            "metadata_size": self.metadata_size
        }
        if self.field is not None:
            result["field"] = self.field
        return result

class RedisResponse:
    """Represents a complete Redis response with metadata and data."""
    
    def __init__(self, data: bytes, metadata: RedisMetadata):
        self.data = data
        self.metadata = metadata
    
    @property
    def data_str(self) -> str:
        """Get data as UTF-8 string."""
        return self.data.decode('utf-8', errors='replace')
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert entire response to dictionary."""
        return {
            "data": self.data_str,
            "metadata": self.metadata.to_dict()
        }

def parse_redis_metadata_buffer(buffer: bytes) -> Union[RedisResponse, bytes]:
    """
    Parse a Redis response buffer that may contain metadata.
    
    Args:
        buffer: Raw bytes from Redis response
        
    Returns:
        RedisResponse object if metadata is present, otherwise raw bytes
        
    Raises:
        ValueError: If buffer format is invalid
    """
    if len(buffer) < 4:
        return buffer  # Too small to contain metadata
    
    # Check for magic header
    if buffer[:4] != b"RDMT":
        return buffer  # Not a metadata buffer, return raw data
    
    if len(buffer) < 20:  # Minimum size for header
        raise ValueError("Invalid metadata buffer: too small")
    
    try:
        # Parse header
        magic = buffer[0:4]
        total_size = struct.unpack("!I", buffer[4:8])[0]
        metadata_size = struct.unpack("!I", buffer[8:12])[0]
        data_size = struct.unpack("!I", buffer[12:16])[0]
        
        if total_size != len(buffer):
            raise ValueError(f"Buffer size mismatch: expected {total_size}, got {len(buffer)}")
        
        offset = 16
        
        # Parse key length and field length (if hash)
        key_len = struct.unpack("!I", buffer[offset:offset+4])[0]
        offset += 4
        
        # Check data type first to determine if we have field length
        # We need to peek ahead to get the data type
        if offset + 12 > len(buffer):  # Need at least 8 more bytes for other fields + 1 for type
            raise ValueError("Buffer truncated")
        
        # For now, assume we might have field length (hash format)
        # We'll determine this by checking remaining space
        remaining_before_strings = metadata_size - offset
        min_remaining = 4 + 8 + 8 + 8 + 1  # data_len + timestamp + hits + misses + type
        
        field_len = 0
        if remaining_before_strings >= min_remaining + 4:  # Extra 4 bytes suggests field length
            field_len = struct.unpack("!I", buffer[offset:offset+4])[0]
            offset += 4
        
        # Parse data length
        data_len = struct.unpack("!I", buffer[offset:offset+4])[0]
        offset += 4
        
        if data_len != data_size:
            raise ValueError(f"Data size mismatch: header says {data_size}, field says {data_len}")
        
        # Parse timestamp and stats
        access_timestamp = struct.unpack("!Q", buffer[offset:offset+8])[0]
        offset += 8
        
        keyspace_hits = struct.unpack("!Q", buffer[offset:offset+8])[0]
        offset += 8
        
        keyspace_misses = struct.unpack("!Q", buffer[offset:offset+8])[0]
        offset += 8
        
        # Parse data type
        data_type = struct.unpack("!B", buffer[offset:offset+1])[0]
        offset += 1
        
        # Parse key string
        if offset + key_len > len(buffer):
            raise ValueError("Buffer truncated while reading key")
        key = buffer[offset:offset+key_len].decode('utf-8')
        offset += key_len
        
        # Parse field string (if hash)
        field = None
        if field_len > 0:
            if offset + field_len > len(buffer):
                raise ValueError("Buffer truncated while reading field")
            field = buffer[offset:offset+field_len].decode('utf-8')
            offset += field_len
        
        # Parse actual data
        if offset + data_len > len(buffer):
            raise ValueError("Buffer truncated while reading data")
        data = buffer[offset:offset+data_len]
        
        # Create metadata object
        metadata = RedisMetadata()
        metadata.key = key
        metadata.field = field
        metadata.data_type = data_type
        metadata.data_size = data_len
        metadata.access_timestamp = access_timestamp
        metadata.keyspace_hits = keyspace_hits
        metadata.keyspace_misses = keyspace_misses
        metadata.total_size = total_size
        metadata.metadata_size = metadata_size
        
        return RedisResponse(data, metadata)
        
    except struct.error as e:
        raise ValueError(f"Invalid buffer format: {e}")
    except UnicodeDecodeError as e:
        raise ValueError(f"Invalid string encoding: {e}")

def redis_get_with_metadata(redis_client, key: str) -> Union[RedisResponse, str]:
    """
    Helper function to perform GET with automatic metadata parsing.
    
    Args:
        redis_client: Redis client instance
        key: Key to retrieve
        
    Returns:
        RedisResponse if metadata enabled, string value otherwise
    """
    result = redis_client.get(key)
    
    if isinstance(result, bytes):
        parsed = parse_redis_metadata_buffer(result)
        if isinstance(parsed, RedisResponse):
            return parsed
    
    return result

def redis_hget_with_metadata(redis_client, key: str, field: str) -> Union[RedisResponse, str]:
    """
    Helper function to perform HGET with automatic metadata parsing.
    
    Args:
        redis_client: Redis client instance
        key: Hash key
        field: Hash field
        
    Returns:
        RedisResponse if metadata enabled, string value otherwise
    """
    result = redis_client.hget(key, field)
    
    if isinstance(result, bytes):
        parsed = parse_redis_metadata_buffer(result)
        if isinstance(parsed, RedisResponse):
            return parsed
    
    return result

# Example usage
if __name__ == "__main__":
    import redis
    
    # Example buffer parsing
    print("=== Redis Metadata Buffer Parser Example ===\n")
    
    try:
        # Connect to Redis
        r = redis.Redis(host='localhost', port=6379)
        
        print("1. Testing without metadata:")
        r.set("test_key", "test_value")
        result = redis_get_with_metadata(r, "test_key")
        print(f"Result: {result}")
        print(f"Type: {type(result)}\n")
        
        print("2. Enabling metadata:")
        r.config_set("read-response-metadata", "yes")
        
        print("3. Testing with metadata:")
        result = redis_get_with_metadata(r, "test_key")
        print(f"Result type: {type(result)}")
        if isinstance(result, RedisResponse):
            print(f"Data: {result.data_str}")
            print(f"Metadata: {result.metadata.to_dict()}")
        else:
            print(f"Raw result: {result}")
        
        print("\n4. Testing hash with metadata:")
        r.hset("test_hash", "field1", "value1")
        result = redis_hget_with_metadata(r, "test_hash", "field1")
        if isinstance(result, RedisResponse):
            print(f"Data: {result.data_str}")
            print(f"Metadata: {result.metadata.to_dict()}")
        else:
            print(f"Raw result: {result}")
            
    except Exception as e:
        print(f"Error: {e}")
        
    print("\n=== Done ===") 